src/rebase.js

```diff
@@ -116,13 +116,22 @@ module.exports = (function(){
   function _fetch(endpoint, options){
     _validateEndpoint(endpoint);
     optionValidators.context(options);
-    optionValidators.then(options);
     options.queries && optionValidators.query(options);
     var ref = firebase.database().ref(endpoint);
     ref = _addQueries(ref, options.queries);
-    ref.once('value', (snapshot) => {
+    return ref.once('value').then(snapshot => {
       var data = options.asArray === true ? _toArray(snapshot) : snapshot.val();
-      options.then.call(options.context, data);
+      if(options.then){
+        options.then.call(options.context, data);
+      }
+      return data;
+    }, err => {
+      //call onFailure callback if it exists otherwise return a rejected promise
+      if(options.onFailure && typeof options.onFailure === 'function'){
+        options.onFailure.call(options.context, err);
+      } else {
```

Comment:
Yes, it is a proper behaviour: no matter what you do inside `onFailure` callback, `_fetch` returns rejected Promise because the fetch operation did fail. As well as it returns resolved Promise when the operation succeeds no matter what happens inside `then` callback. That is `fetch({then:onSuccess, onFailure: onFailure})` ignores values returned from callbacks (if any), while `fetch({}).then(onSuccess, onFailure)` respects those returned values and pass them down to chained `then`s and `catch`es. I see this as the expected behaviour for those who prefer callbacks, and for those who use Promises as well.

Thus, I suppose the only fix needed is to remove this `else` on line 132.

What do you think?